pkg/R/utils.R

```diff
@@ -164,11 +176,14 @@ reserialize <- function(rdd) {
   if (!inherits(rdd, "RDD")) {
     stop("Argument 'rdd' is not an RDD type.")
   }
-  if (rdd@env$serialized) {
-    return(rdd)
-  } else {
+  if (rdd@env$serializedMode != "byte") {
+    if (rdd@env$serializedMode == "row") {
+      rdd@env$serialzedMode == "byte"
```

Comment:
Agreed on the invariants. Not sure on the best way to set hard categories for function arguments in R (as opposed to the soft categories imposed by the checks in RDD's initialize function now).

For digging into the reserialize issue, I'd suggest starting with test_sparksql.R as it has tests for unionRDD and objectFile that should provide a good example of how the serializedMode is handled in different cases.